### PR TITLE
Download Data onto User Device

### LINKED ISSUE #6: Download Data onto User Device
Start by creating a new directory in the frontend folder that is responsible for interacting with the device's file system. For this task, implement the method that fetch the data from MongoDB and downloads it onto the user's device. First priority is to just make a working method that does this, then if time, implement logic to check if the user's device has the latest version and needs to fetch new data (may need to add version number to mongo database). 

If a user has an outdated version, the current info stored on their device should be wiped, and all the documents should be download from mongodb onto their device again. This makes updates simpler, as rather than checking which documents have been deleted/added/newly added, can wipe and readd everything. 

Helpful documentation: https://docs.expo.dev/versions/latest/sdk/filesystem/

### PR DESCRIPTION
**Pull request by LillianHo5**
## Tracking Info

Resolves #6 

## Changes

Created the functionality to fetch data from the MongoDB database and download it onto the user's device. A download occurs whenever the user opens the app and is connected to the internet. Implemented logic to wipe previously downloaded data from the device if the user has an outdated version, replacing it with new data from the latest version. 

## Testing

We tested our changes on physical devices and emulators for IOS and Android. To ensure each component worked, we added a collection (versions) to the RohanTest database and had a document indicating the current version of the data. We had several log statements to print the results of each download made, testing several cases, including: 

- User does not currently have any data downloaded on their device (newest data is downloaded)
- User has an outdated version on their device (old data is wiped and the newest data is downloaded)
- The latest version and version on the user's device matches (no data is fetched/no download occurs) 
- User opens the application and is connected to the internet (download should occur if the first 2 cases occur) 

## Confirmation of Change

When testing on a physical device, make sure to create a `.env` file in the `dfm-sideline-sidekickapp` directory using the `.env.sample` format. Insert your IP address as the last environment variable. 
- Windows:
    - To retrieve your IP address, go to your command line/terminal and type `ipconfig`. It is under `IPv4` address.
- Mac/Linux 
    - To retrieve your IP address, go to your command line/terminal and type `ifconfig en0`. It is under `inet`.

- example: `PUBLIC_EXPO_IP_ADDRESS=123.123.123.12`
    - NO quotation marks around the IP address.

Edit: In the `.env`, make sure it is modified to point to the RohanTest database. 
- example format: `OUR_DATABASE_URL/RohanTest?retryWrites=true&w=majority`

Example output: 
image

### CODE REVIEW

File: dfm-sideline-sidekick-app/download/downloadFromAPI.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+  // checking if file exists, will be null if not
+  const fileExists = await checkFileExists(fileDir, fileName);
+
+  try {
+    // will store the path to our file
+    let uri = "";
+
+    const newestVersion = (await getCurrentVersion(versionUrl))[0].version as string;
+    const storedVersion = await getStoredVersion();
```
Comment (Anthonyp0329) [inline dfm-sideline-sidekick-app/download/downloadFromAPI.ts:41]: Here, when it calls getStoredVersion, it errors if a version isn't already stored, so need to add some logic either here or in the getStoredVersion that acounts for the first time this version compare is being done, and no version is stored on the user device yet. My understanding could also be flawed, so lmk if you guys can reproduce this issue.
Here's the error I get:
image

File: dfm-sideline-sidekick-app/download/downloadFromAPI.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import { getStoredVersion, setStoredVersion } from "./versionControl/storedVersion";
+
+export const downloadJSON = async (fileName: string, OS: string) => {
+  let localhost;
+
+  // check if currently running on an emulator or device
+  const isDevice = checkDevice();
+
+  if (isDevice) {
+    localhost = "100.115.33.192"; // ip address of laptop (for running on physical devices) [1. go to command line, 2. type ipconfig /all 3. it's under IPv4 address]
```
Comment (Anthonyp0329) [inline dfm-sideline-sidekick-app/download/downloadFromAPI.ts]: For the IP addresses, maybe make a .env file with a key to store its more easy/transparent for others to quickly swap

**Review by Anthonyp0329 — CHANGES_REQUESTED** [review]
Holy crap guys, amazing amazing work! Just a few changes - I found an issue where the program errors if a version isn't already stored. Other than this, looks great!

File: dfm-sideline-sidekick-app/download/downloadFromAPI.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    console.log(); // feel free to remove these extra logs, they're just for output clarity when debugging
+
+    console.log("PRINT VERSIONS BEFORE UPDATE");
+    console.log("UPDATED DEVICE VERSION:", storedVersion);
+    console.log("NEWEST VERSION:", newestVersion);
+
+    console.log();
+
+    // no stored version or current doesn't match stored version
+    if (!fileExists || !storedVersion || storedVersion < newestVersion) {
```
Comment (Anthonyp0329) [inline dfm-sideline-sidekick-app/download/downloadFromAPI.ts]: change the < to a != in case we switch up version number conventions ever

**Review by Anthonyp0329 — APPROVED** [review]
Looks good! Just had to make sure I had right iP address and used correct connection string. also make sure to resolve merge conflicts.

File: dfm-sideline-sidekick-app/download/downloadFromAPI.ts (hunk 2 above)
Comment (Anthonyp0329) [reply dfm-sideline-sidekick-app/download/downloadFromAPI.ts]: (can use the backend one for reference, and make sure to make a .env.sample but don't commit the actual .env)

Comment (EdwardLinS) [reply dfm-sideline-sidekick-app/download/downloadFromAPI.ts]: added!

File: dfm-sideline-sidekick-app/download/downloadFromAPI.ts (hunk 3 above)
Comment (EdwardLinS) [reply dfm-sideline-sidekick-app/download/downloadFromAPI.ts]: changed!

**Review by PiXlRAM — APPROVED** [review]
LGTM, worked perfectly when tested on an android virtual device.

**Review by r800360 — APPROVED** [review]
Lillian and Edward, excellent work with these changes to incorporate the downloading of data onto the user device from the MongoDB database. The code changes look so clean, and I was able to get this to work when testing with my Android phone and my iPad. I went ahead and fixed merge conflicts with the main branch to prepare this code for merging. Nice job!

**Pull request merged**